### ISSUE TITLE
SelectOnStep::onKey returns wrong type, preventing combinations of join predicates

### ISSUE THREAD

**Issue by SebastianS90** [jOOQ/jOOQ#12200]
### Use case:

```sql
CREATE TABLE exams (
    id INT NOT NULL PRIMARY KEY
    -- additional data (title, ...) goes here
)

CREATE TABLE examresults (
    exam_id INT NOT NULL REFERENCES exams(id),
    student_id INT NOT NULL,
    grade INT NOT NULL,
    PRIMARY KEY (exam_id, student_id)
)
```

I want to see all exams and additionally the grade I (student_id 42) have achieved there if there is one:
```java
ctx.select(EXAMS.asterisk(), EXAMRESULTS.GRADE)
    .from(EXAMS)
    .leftJoin(EXAMRESULTS).on(EXAMRESULTS.EXAM_ID.eq(EXAMS.ID)).and(EXAMRESULTS.STUDENT_ID.eq(42))
```
But without explicitly stating the foreign key connection in the join condition.

Code repo: https://github.com/SebastianS90/jooq-onkey-condition

### Possible solution you'd like to see:

```java
    .leftJoin(EXAMRESULTS).onKey().and(EXAMRESULTS.STUDENT_ID.eq(42))  // does not compile
```

### Possible workarounds:

The original code above stating the explicit join condition.

Note that we cannot use
```java
    .leftJoin(EXAMRESULTS).onKey()
    .where(EXAMRESULTS.STUDENT_ID.eq(42))
```
because this one has a different semantics.

### Versions:

- jOOQ: 3.15.1
- Java: 11

**Comment by lukaseder** [comment]
Thanks for your suggestion. That makes sense, and it's possible with this syntax:

```java
ctx.select(EXAMS.asterisk(), EXAMRESULTS.GRADE)
    .from(EXAMS
      .leftJoin(EXAMRESULTS).onKey().and(EXAMRESULTS.STUDENT_ID.eq(42)))
    .fetch();
```

I.e. `TableOnStep::onKey` returns the expected `TableOnConditionStep`. But the return type of the "join convenience syntax" `SelectOnStep::onKey` is wrong

**Comment by lukaseder** [comment]
Since this change produces a binary incompatible API change, I won't backport the fix to 3.15.2, but will definitely fix for 3.16.0 right away.
